test/skate_web/controllers/detour_route_controller_test.exs

```diff
@@ -12,7 +12,33 @@ defmodule SkateWeb.DetourRouteControllerTest do
 
   describe "directions" do
     defp directions_json(coordinates: coordinates) do
-      %{"features" => [%{"geometry" => %{"coordinates" => coordinates}}]}
+      %{
+        "features" => [
+          %{
+            "geometry" => %{"coordinates" => coordinates},
+            "properties" => %{
+              "segments" =>
+                case length(coordinates) do
+                  0..1 ->
+                    []
+
+                  len ->
+                    Enum.map(1..(len - 1), fn index ->
+                      %{
+                        "steps" => [
+                          %{
+                            "instruction" => "Turn right onto #{index}",
+                            "name" => to_string(index),
+                            "type" => 1
+                          }
+                        ]
+                      }
+                    end)
+                end
+            }
+          }
+        ]
+      }
```

Comment:
Sort of the same comment that I made [earlier](https://github.com/mbta/skate/pull/2367/files#r1458081312)... I'd rather have the test data be more explicit. For instance, 

```suggestion
      %{
        "features" => [
          %{
            "geometry" => %{"coordinates" => coordinates},
            "properties" => %{
              "segments" => turn_by_turn_segments,
            }
          }
        ]
      }
```

Where `turn_by_turn_segments` is an argument passed in like `coordinates` (Github won't let me suggest a change to a line you didn't edit 😢).

Then `turn_by_turn_segments` can be passed in as part of the setup of each test (or given a sane default for any test that doesn't care about the turn-by-turns ((I think in order to do sane defaults, we'll need to torture Elixir a bit - keyword args with default values is not one of Elixir's strong core competencies))).

As an added bonus, we could make the test `"returns directions as a flat list"` below a lot clearer, since right now, it's dealing with a coordinate array, rather than the nested `segment` structure.

Reply:
(posting for discussion)

I'd also like to try using factories for this, and have an example implementation in elixir to review
```elixir

  def ors_directions_step_json_factory do
    %{
      "instruction" => sequence("ors_instruction_step"),
      "name" => sequence("ors_instruction_name"),
      "type" => sequence("ors_instruction_type", [0, 1, 2, 3, 4, 5, 6, 7, 12, 13])
    }
  end

  def ors_directions_segment_json_factory do
    %{
      "steps" =>
        build_list(
          sequence("ors_segment_json_num_steps", [4, 1, 3, 2]),
          :ors_directions_step_json
        )
    }
  end

  def ors_directions_json_factory(attrs) do
    coordinates = Map.get(attrs, :coordinates, [[0, 0], [1, 1], [2, 2]])

    segments =
      Map.get(attrs, :segments, build_list(3, :ors_segment_json))

    %{
      "features" => [
        %{
          "geometry" => %{"coordinates" => coordinates},
          "properties" => %{
            "segments" => segments
          }
        }
      ]
    }
    |> evaluate_lazy_attributes()
  end
```

Usage of this is would/could be along the lines of
```elixir
Skate.Factory.build(:ors_directions_json, coordinates: [...])
Skate.Factory.build(:ors_directions_json, %{coordinates: [...]})
Skate.Factory.build(:ors_directions_json, segments: [...], coordinates: [...])
Skate.Factory.build(:ors_directions_json, segments: [...])

Skate.Factory.build(:ors_directions_json, 
	segments: [
		Skate.Factory.build(:ors_directions_segment_json, 
			%{"steps" => Skate.Factory.build_list(2, :ors_directions_step_json, 
				[{"type", 1}])}),
		Skate.Factory.build(:ors_directions_segment_json, 
			%{"steps" => Skate.Factory.build_list(2, :ors_directions_step_json, 
				%{"type" => 0, "name" => "test-name"})})
	]
)

Reply:
Definitely interested in pursuing this. I'm interested in learning more about the advantages that this has over a slightly expanded `directions_json` (expanded to add `segments` as a parameter).

I assume that we wouldn't want to support both `%{"type" => 0, "name" => "test-name"}` format and `[{"type", 1}]`? Or is there some zaniness in ORS's API that I'm not thinking of?

Reply:
The library [ex_machina](https://hexdocs.pm/ex_machina/readme.html) which provides the `Skate.Factory.build` family of functions takes either a keyword list, or a map as input. I think behind the scenes it collapses them both into a map, but the syntax here is just the different ways that the `build` function _can_ be used, all are valid expressions (y'know, if the `[...]` were valid data).

The `[{"type", 1}]` is just required because it's different from `[type: 1]` and `["type": 1]` is not a valid expression, so instead, you make a tuple of the key name as a string, and the value.

Reply:
PTALA [`1abe9cc` (#2367)](https://github.com/mbta/skate/pull/2367/commits/1abe9cc486b7ee4df3bd457f91c7b5b4171f6440) 🙏🏻